pysnark/runtime.py

```diff
@@ -204,97 +204,203 @@ def __ge__(self, other): return (self-other).check_positive()
     def assert_ge(self, other, err=None): (self-other).assert_positive(err)
         
     def __bool__(self):
-        raise NotImplementedError("Should not run bool() on a LinComb")
-        
+        raise NotImplementedError("Cannot call __bool__ on a LinComb. \
+            Instead of if statements, use if_then_else from pysnark.branching")
+
     def __add__(self, other):
+        """
+        Adds a LinComb with an integer or another LinComb
+        Costs 0 constraints
+        """
+        if isinstance(other, int):
+            return self + ConstVal(other)
         if isinstance(other, LinComb):
-            return LinComb(self.value+other.value, self.lc+other.lc)
-        elif is_base_value(other):
-            return self + other*LinComb.ONE
-        else:
-            return NotImplemented
-        
-    def __sub__(self, other):
-        return self+(-other)
+            return LinComb(self.value + other.value, self.lc + other.lc)
+        return NotImplemented
     
+    def __sub__(self, other):
+        """
+        Subtracts an integer or another LinComb from a LinComb
+        Costs 0 constraints
+        """
+        return self + (-other)
+
     def __mul__(self, other):
+        """
+        Multiplies a LinComb with an integer or another LinComb
+        Costs 0 constraints to multiply with an integer
+        Costs 1 constraint to multiply with a LinComb
+        """
+        if isinstance(other, int):
+            return LinComb(self.value * other, self.lc * other)
         if isinstance(other, LinComb):
-            retval = PrivVal(self.value*other.value)
+            retval = PrivVal(self.value * other.value)
             add_constraint_unsafe(self, other, retval)
             return retval
-        elif is_base_value(other):
-            return LinComb(self.value*other, self.lc*other)
-        else:
-            return NotImplemented
-                    
-    def __matmul__(self, other): return NotImplemented
-    
+        return NotImplemented
+
     def __truediv__(self, other): 
-        """ Proper division """
+        """
+        Divides a LinComb with an integer or another LinComb using integer division
+        Throws ValueError if values are not evenly divisible
+        Costs 0 constraints to divide with an integer
+        Costs 1 constraint to divide with a LinComb
+        """
+        if isinstance(other, int):
+            if other == 0:
+                raise ValueError("Division by zero")
+            if is_guard() and (self.value % other == 0):
+                return LinComb(self.value // other, self.lc * backend.fieldinverse(other))
+            if ignore_errors():
+                return LinComb.ZERO
```

Comment:
This seems to result in different constraints depending on whether self.value % other == 0. I think instead of LinComb.ZERO, LinComb(0, self.lc * backend.fieldinverse(other)) should be returned